Add libcvmfs calls that list and stat nested catalogs

nhazekam opened the pull request:
Added calls for getting information about nested catalogs using libcvmfs. Also added were several tests for the new calls.

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+  shash::Any hash;
+  uint64_t size;
+  const bool found =
+    mount_point_->catalog_mgr()->GetRootCatalog()->FindNested(p, &hash, &size);
```
Let's discuss this once more.  I think `FindNested` alone is not sufficient, as the correct nested catalog might not be attached yet.  Being a public method now, it would also need to apply the appropriate lock (`ReadLock` or `WriteLock`). I think we need to wrap the `FindNested` call after all, in a public `LookupNested` method.

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+  }
+
+  /* Set values of the passed structure */
+  info->mountpoint = c_path;
```
That could be a parent path of c_path, I think.

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+
+  /* If dirent exists and is a nested catalog mountpoint, 
+   * we look past it so the catalog is mounted. */
+  if ( dirent.IsNestedCatalogMountpoint() ) {
```
no spaces inside parantheses

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+    /* Look at fake file */
+    const std::string fake("/.cvmfscatalog");
+    std::string extended_path = path.ToString() + fake;
+    path.Assign(extended_path.c_str(), strlen(extended_path.c_str()));
```
Simpler: `path.Append(...)`

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+  AppendStringToList(NULL, buf, &listlen, buflen);
+
+  /* Build listing */
+  catalog::Catalog *parent  = found_catalog->parent();
```
Hm, for looking reasons, we might need to wrap the core of this call is well in the catalog manager, perhaps as `ListNestedCatalogSkein`

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+
+  /* Build listing */
+  catalog::Catalog *parent  = found_catalog->parent();
+  if ( parent ) {
```
white spaces

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+  if ( parent ) {
+    /* Walk up parent tree to find base */
+    std::vector<catalog::Catalog*> parents;
+    while ( parent->HasParent() ) {
```
white spaces

jblomer commented on cvmfs/libcvmfs_int.cc:
```diff
@@ -374,6 +374,107 @@ int LibContext::ListDirectory(
+      parent = parent->parent();
+    }
+    parents.push_back(parent);
+    while ( !parents.empty() ) {
```
white spaces

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -238,3 +251,217 @@ TEST_F(T_Libcvmfs, Attachv2) {
+  // The buf length should be at least 3
+  EXPECT_LE(3, buflen);
+  // Check that listed info matches specified dir
+  EXPECT_TRUE(!strcmp(".",  buf[0]));
```
Perhaps easier as `EXPECT_FALSE(strcmp(...))`

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -238,3 +251,217 @@ TEST_F(T_Libcvmfs, Attachv2) {
+  cvmfs_options_fini(opts);
+}
+
+TEST_F(T_Libcvmfs, StatNestedCatalog) {
```
This should test finding deeply nested catalogs, too.  E.g. /nested/sub-nested in the case not even /nested has been attached yet.

nhazekam commented on cvmfs/catalog.cc:
```diff
@@ -676,6 +676,44 @@ const Catalog::NestedCatalogList Catalog::ListOwnNestedCatalogs() const {
+ * @return  a list of all nested catalogs.
+ */
+const Catalog::NestedCatalogNameList
+  Catalog::ListOwnNestedCatalogsSkein() const {
```
How would you like this line formatted?

nhazekam commented:
@jblomer Please take a look when you get the chance.

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -63,10 +69,16 @@ class T_Libcvmfs : public ::testing::Test {
   string opt_cache_;
   FILE *fdevnull_;
+
+  int fd_cwd_;
```
Are this and the next one used?

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -238,3 +251,264 @@ TEST_F(T_Libcvmfs, Attachv2) {
+  EXPECT_FALSE(buf[4]);
+  cvmfs_list_destroy(buf);
+
+  /* Finalize and close repo and options */
```
I think there is a `cvmfs_detach_repo` missing

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -238,3 +251,264 @@ TEST_F(T_Libcvmfs, Attachv2) {
+
+
+  /* Finalize and close repo and options */
+  cvmfs_fini();
```
+ `cvmfs_detach_repo`

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -238,3 +251,264 @@ TEST_F(T_Libcvmfs, Attachv2) {
+  cvmfs_list_destroy(buf);
+
+  /* Finalize and close repo and options */
+  cvmfs_fini();
```
+ `cvmfs_detach_repo`

jblomer commented on cvmfs/catalog_mgr_impl.h:
```diff
@@ -294,6 +295,181 @@ bool AbstractCatalogManager<CatalogT>::LookupPath(const PathString &path,
+    }
+  }
+
+  result = GetRootCatalog()->FindNested(catalog->root_prefix(), hash, size);
```
Which is strange... shouldn't the unit tests have found it?

nhazekam commented:
@jblomer I think I have addressed all of the comments from the previous commit, as well as rebasing to resolve the conflicting test. Please take a look.

nhazekam commented:
@jblomer The recent changes now detect the difference between using the correct catalog to find the Nested Catalog versus just using the Root catalog. I tested the incorrect way and the test does fail.
